### PR TITLE
Add `"branch"` scope type

### PR DESCRIPTION
**Pull request by pokey**
- Partially addresses https://github.com/cursorless-dev/cursorless/issues/1173
- Split from https://github.com/cursorless-dev/cursorless/pull/831

## Checklist

|             | Python | Rust | Java | Typescript | C++ | C#  |
| ----------- | ------ | ---- | ---- | ---------- | --- | --- |
| Switch-case | ✅     | ✅   | ✅   | ✅         |     |     |
| If-else     | ✅     | ✅   |      | ✅         |     |     |
| Try-catch   | ✅     |   N/A   |      |     ✅       |     |     |
| Ternary     |   ✅     | ✅   |      |     ✅       |     |     |

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [x] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [x] I have not broken the cheatsheet

Co-authored-by: Michael Doronin <warrior2031@mail.ru>

### CODE REVIEW

**Comment by pokey** [comment]
Closing until someone has bandwidth to finish this PR

**Comment by pokey** [comment]
After discussion in meet-up:

Just kidding; let's get this thing in for the languages that exist here.  more useful to at least have it functional in some places

File: src/test/suite/fixtures/recorded/languages/python/clearBranch14.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      pass
+    else:
+      pass
+  selections:
+    - anchor: {line: 0, character: 0}
+      active: {line: 0, character: 0}
+  marks: {}
+finalState:
+  documentContents: |-
+
```
Comment (auscompgeek) [inline src/test/suite/fixtures/recorded/languages/python/clearBranch14.yml:26]: It feels a bit weird for `branch` to work on the `try` block itself 🤷

**Comment by auscompgeek** [comment]
We should also test:

- [ ] Python for/else and while/else
- [ ] Rust `if let` and let/else

File: src/test/suite/fixtures/recorded/languages/python/clearBranch14.yml (hunk 1 above)
Comment (pokey) [reply src/test/suite/fixtures/recorded/languages/python/clearBranch14.yml:26]: Yeah I'm not entirely sure what the use case for that is tbh 😅, but seemed consistent, and maybe someone would want to use it with the `copy` action?

**Comment by auscompgeek** [comment]
> Unfortunately that is blocked on #1201 😕

Ah, I was afraid as such.

File: cursorless-talon/src/modifiers/scopes.py [hunk 2]
```diff
@@ -46,6 +46,7 @@
     "selector": "selector",
     "state": "statement",
     "string": "string",
+    "branch": "branch",
```
Comment (auscompgeek) [inline cursorless-talon/src/modifiers/scopes.py:49]: nit: any reason this is listed in this order in particular?

**Review by auscompgeek — APPROVED** [review]
I think the implicit "this" change got cherry-picked into this branch for some reason, but otherwise LGTM.

I presume it should be trivial to reuse the if/else matchers for Java and C++ like we did for Rust, and I hope the same for try/catch, so might be worth doing it here anyway.

**Comment by pokey** [comment]
> I think the implicit "this" change got cherry-picked into this branch for some reason, but otherwise LGTM.

Strange.  That's already merged into main anyway, but looks like the parentage is strange here; I'll have a look

> I presume it should be trivial to reuse the if/else matchers for Java and C++ like we did for Rust, and I hope the same for try/catch, so might be worth doing it here anyway.

Unfortunately not 😕.  Java, C#, and C++ don't have a special node for else clauses, so the logic gets more complicated.  Should be doable, but not just a quick afterthought like one would hope.  I'd be inclined to leave for follow-up work

File: cursorless-talon/src/modifiers/scopes.py (hunk 2 above)
Comment (pokey) [reply cursorless-talon/src/modifiers/scopes.py:49]: Not really.  I should prob file a follow-up to alphabetise; didn't want to mess with the diff

**Review by auscompgeek — APPROVED** [review]

**Pull request merged**